File: behavioral/chain-of-responsibility.php
<?php
abstract class Handler {
    protected ?Handler $next = null;
    public function setNext(Handler $h){ $this->next = $h; return $h; }
    abstract public function handle(int $amount);
}
class LowHandler extends Handler {
    public function handle(int $amount){ 
        if($amount <= 100) echo "Low handled\n";
        elseif($this->next) $this->next->handle($amount);
    }
}
class HighHandler extends Handler {
    public function handle(int $amount){
        if($amount > 100) echo "High handled\n";
        elseif($this->next) $this->next->handle($amount);
    }
}

$h1 = new LowHandler();
$h2 = new HighHandler();
$h1->setNext($h2);
$h1->handle(150);

File: behavioral/chain-of-responsibility.go
package main

import "fmt"

type Handler interface {
	SetNext(Handler) Handler
	Handle(int)
}

type Base struct{ next Handler }

func (b *Base) SetNext(n Handler) Handler { b.next = n; return n }

type LowHandler struct{ Base }

func (h *LowHandler) Handle(a int) {
	if a <= 100 {
		fmt.Println("Low handled")
	}
	if h.next != nil {
		h.next.Handle(a)
	}
}

type HighHandler struct{ Base }

func (h *HighHandler) Handle(a int) {
	if a > 100 {
		fmt.Println("High handled")
	}
	if h.next != nil {
		h.next.Handle(a)
	}
}

func main() {
	h1 := &LowHandler{}
	h2 := &HighHandler{}
	h1.SetNext(h2)
	h1.Handle(150)
}
